File: old-code/routes/abilities.py
import random
from flask import render_template, redirect, url_for, request
from routes import abilities_bp
from lib import get_player
from lib.turns import TurnManager
from lib.database import db


@abilities_bp.route('/')
def index():
    player = get_player()
    success = request.args.get('success', 'false').lower() == 'true'
    return render_template(
        'pages/../templates/pages/abilities.html', player_name=player.name, abilities=player.abilities.values(),
        turns=TurnManager.turns, **{**request.args, 'success': success})


@abilities_bp.route('/check', methods=['POST'])
def check():
    player = get_player()
    ability_name = request.form['ability_name']
    ability = player.abilities[ability_name]
    target_type = request.form['target_type']
    if target_type == 'difficulty_level':
        target_check = int(request.form['difficulty_level'])
    elif target_type == 'character':
        character_index = int(request.form['character_index'])
        # noinspection PyUnresolvedReferences
        character: TurnManager.Turn = TurnManager.turns[character_index]
        target_ability = character.abilities[request.form['target_ability']]
        target_check = target_ability.level + target_ability.tmp_diff + random.randint(1, 20)
    else:
        return render_template('pages/../templates/pages/error.html', message=f'INVALID TARGET TYPE: {target_type}')
    ability_check = random.randint(1, 20)
    total = ability_check + ability.level + ability.tmp_diff
    success = ability_check >= target_check
    db.ability_check(success, player.name, ability_name)
    return redirect(url_for(
        'abilities.index', ability_name=ability_name, ability_level=ability.level, ability_check=ability_check,
        check_total=total, target_check=target_check, tmp_diff=ability.tmp_diff, success=success))

File: backend/src/routes/abilities.js
import {Router} from 'express'
import {getPlayer, deleteTemplate, processSliderValues} from '../lib/index.js'
import Database from '../lib/database.js'
import {CharacterType, AbilityCheckTargetType} from '../../../shared.js'
import {randomInt} from 'crypto'

const abilitiesRoutes = Router()

const getRandomInt = (min, max) => randomInt(max - min + 1) + min

abilitiesRoutes.post('/check', async (req, res) => {
    const {
        abilityName,
        targetType,
        characterType,
        npcName
    } = req.body
    let character
    if (characterType === CharacterType.Player) {
        character = await getPlayer(req)
        if (!character) return res.status(400).send('Player not found')
    } else {
        character = await Database.getNpc(npcName)
    }
    const findAbility = (abilities, name) => abilities.find(ability => ability.name === name)
    const ability = findAbility(character.abilities, abilityName)
    if (!ability) return res.status(400).send('Ability not found')
    let targetCheck
    if (targetType === AbilityCheckTargetType.DifficultyLevel) {
        const {difficultyLevel} = req.body
        targetCheck = difficultyLevel
    } else if (targetType === AbilityCheckTargetType.Character) {
        const {
            targetName,
            targetCharacterType,
            targetAbilityName
        } = req.body
        let targetCharacter
        if (targetCharacterType === CharacterType.Player) {
            targetCharacter = await Database.getPlayer(targetName)
        } else {
            targetCharacter = await Database.getNpc(targetName)
        }
        if (!targetCharacter) return res.status(400).send('Character not found')
        const characterAbility = findAbility(targetCharacter.abilities, targetAbilityName) || {level: 0, tmpDiff: 0}
        if (!characterAbility) return res.status(400).send('Character ability not found')
        targetCheck = characterAbility.level + characterAbility.tmpDiff + getRandomInt(1, 20)
    } else {
        return res.status(400).send('Invalid ability check target type')
    }
    const check = getRandomInt(1, 20)
    const total = check + ability.level + ability.tmpDiff
    const success = check >= targetCheck
    await Database.abilityCheck(success, character, ability, characterType)
    return res.status(201).json({
        success,
        abilityName,
        check,
        total,
        targetCheck
    })
})

abilitiesRoutes.post('/template', async (req, res) => {
    const {
        name,
        ...rest
    } = req.body
    const effectedConditions = processSliderValues(rest, 'ability')
    try {
        const template = {
            name,
            effectedConditions
        }
        const templates = await Database.addAbilityTemplate(template)
        res.status(201).json({templates})
    } catch (error) {
        if (error.name === 'PrismaClientKnownRequestError') {
            return res.status(400).send('Ability template already exists')
        }
        return res.status(500).send(error.message)
    }
})

abilitiesRoutes.delete('/template', async (req, res) => {
    return await deleteTemplate(req, res, Database.deleteAbilityTemplate)
})

export default abilitiesRoutes
